File: 26S/session3.py
# session3.py
# Session 3: Subclassification, Regression Foundations, Regression in RCTs
# Three demos:
#   A. Manual subclassification on a binary stratifier.
#   B. Regression as automated subclassification (same answer, less labour).
#   C. Why regression covariates help even in a randomised experiment
#      (variance reduction).
#
# Required packages:
#   pip install causaldata pandas numpy statsmodels

import numpy as np
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf
from causaldata import nsw_mixtape

rng = np.random.default_rng(2026)
df  = nsw_mixtape.load_pandas().data

# ============================================================
# Part A. Manual subclassification on `nodegree`
# ============================================================
strata = (df.groupby("nodegree")
            .apply(lambda g: pd.Series({
                "n_t":   (g.treat == 1).sum(),
                "n_c":   (g.treat == 0).sum(),
                "ybar_t": g.loc[g.treat == 1, "re78"].mean(),
                "ybar_c": g.loc[g.treat == 0, "re78"].mean(),
            }))
            .assign(diff = lambda d: d.ybar_t - d.ybar_c,
                    n    = lambda d: d.n_t + d.n_c))

ate_manual = (strata["diff"] * strata["n"]).sum() / strata["n"].sum()
print("---- Manual subclassification on `nodegree` ----")
print(strata)
print(f"\n  Weighted ATE (manual): {ate_manual:+.1f}")

# ============================================================
# Part B. Same idea via OLS regression
# ============================================================
m_reg = smf.ols("re78 ~ treat + nodegree", data=df).fit()
print(f"\n  Regression coefficient on treat: "
      f"{m_reg.params['treat']:+.1f}  (SE = {m_reg.bse['treat']:.1f})")
# Regression weighting differs slightly from naive stratum-size weighting
# (regression weights by within-stratum variance of T) — close, not identical.

# ============================================================
# Part C. Regression in an RCT — why covariates still help
# ============================================================
n = 500
X = rng.normal(size=n)
T = rng.binomial(1, 0.5, size=n)             # truly random assignment
Y = 2 + 5 * X + 1.0 * T + rng.normal(size=n)  # true ATE = 1.0
sim = pd.DataFrame({"Y": Y, "T": T, "X": X})

m_naive = smf.ols("Y ~ T",     data=sim).fit()
m_adj   = smf.ols("Y ~ T + X", data=sim).fit()

print("\n---- RCT: covariate adjustment for variance reduction ----")
print(f"  Naive RCT estimate:  tau = {m_naive.params['T']:+.3f}  "
      f"(SE = {m_naive.bse['T']:.3f})")
print(f"  With covariate X:    tau = {m_adj.params['T']:+.3f}  "
      f"(SE = {m_adj.bse['T']:.3f})  <- tighter")

File: 26S/session3.R
# session3.R
# Session 3: Subclassification, Regression Foundations, Regression in RCTs
# Three demos:
#   A. Manual subclassification on a binary stratifier.
#   B. Regression as automated subclassification (same answer, less labour).
#   C. Why regression covariates help even in a randomised experiment
#      (variance reduction).
#
# Required packages:
#   install.packages(c("causaldata", "tidyverse"))

library(causaldata)
library(tidyverse)
set.seed(2026)

# ============================================================
# Part A. Manual subclassification on NSW (LaLonde job training)
# ------------------------------------------------------------
# We stratify on `nodegree` (a binary baseline covariate) and compute the
# stratum-weighted ATE. This is the simplest possible adjustment estimator.
# ============================================================
df <- nsw_mixtape

strata <- df %>%
  group_by(nodegree) %>%
  summarise(n_t = sum(treat == 1),
            n_c = sum(treat == 0),
            ybar_t = mean(re78[treat == 1]),
            ybar_c = mean(re78[treat == 0]),
            diff   = ybar_t - ybar_c,
            n_total = n_t + n_c,
            .groups = "drop")

ate_manual <- with(strata, sum(diff * n_total) / sum(n_total))
cat("---- Manual subclassification on `nodegree` ----\n")
print(strata)
cat(sprintf("\n  Weighted ATE (manual): %+.1f\n", ate_manual))

# ============================================================
# Part B. Same idea via OLS regression
#         lm(Y ~ T + nodegree)  recovers the same adjusted effect.
# ============================================================
m_reg <- lm(re78 ~ treat + nodegree, data = df)
cat(sprintf("\n  Regression coefficient on treat: %+.1f  (SE = %.1f)\n",
            coef(m_reg)["treat"], sqrt(diag(vcov(m_reg)))["treat"]))
# Note: regression weighting differs slightly from the simple stratum-size
# weighting (regression weights by within-stratum variance of T) — close,
# not identical. Both target the same identified estimand under ignorability.

# ============================================================
# Part C. Regression in an RCT — why covariates still help
# ------------------------------------------------------------
# Simulate an RCT with a strong predictor X of Y. Treatment is assigned
# randomly, so OLS (Y ~ T) is unbiased — but adding X as a control
# REDUCES the standard error sharply.
# ============================================================
n  <- 500
X  <- rnorm(n)
T  <- rbinom(n, 1, 0.5)                # truly random assignment
Y  <- 2 + 5 * X + 1.0 * T + rnorm(n)   # true ATE = 1.0

m_naive <- lm(Y ~ T)
m_adj   <- lm(Y ~ T + X)

cat("\n---- RCT: covariate adjustment for variance reduction ----\n")
cat(sprintf("  Naive RCT estimate:   tau = %+.3f  (SE = %.3f)\n",
            coef(m_naive)["T"], sqrt(diag(vcov(m_naive)))["T"]))
cat(sprintf("  With covariate X:     tau = %+.3f  (SE = %.3f)  <- tighter\n",
            coef(m_adj)["T"],   sqrt(diag(vcov(m_adj)))["T"]))
